src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,787 @@
+{
+	Datum values[Natts_pg_dist_commit_transaction];
+	bool isNulls[Natts_pg_dist_commit_transaction];
+	uint64 clockLogical = GET_LOGICAL(transactionClockValue);
+	uint32 clockCounter = GET_COUNTER(transactionClockValue);
+
+	ereport(DEBUG1, (errmsg("Persisting transaction %s with "
+							"clock logical (%lu) and counter(%u)",
+							transactionId, clockLogical, clockCounter)));
+
+	/* form new transaction tuple */
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	uint64 transactionTimestamp = GetEpochTimeMs();
+
+	values[Anum_pg_dist_commit_transaction_transactionid - 1] =
+		CStringGetTextDatum(transactionId);
+	values[Anum_pg_dist_commit_transaction_clock_logical - 1] =
+		Int64GetDatum(clockLogical);
+	values[Anum_pg_dist_commit_transaction_clock_counter - 1] =
+		Int64GetDatum(clockCounter);
+	values[Anum_pg_dist_commit_transaction_timestamp - 1] =
+		Int64GetDatum(transactionTimestamp);
+
+	/* open pg_dist_commit_transaction and insert new tuple */
+	Relation pgDistCommitTransaction =
+		table_open(DistCommitTransactionRelationId(), RowExclusiveLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistCommitTransaction);
+	HeapTuple heapTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+
+	CatalogTupleInsert(pgDistCommitTransaction, heapTuple);
+
+	CommandCounterIncrement();
+
+	/* close relation and invalidate previous cache entry */
+	table_close(pgDistCommitTransaction, NoLock);
+}
+
+
+/*
+ * Initialize the shared memory clock value to the highest clock
+ * persisted. This will protect from any clock drifts.
+ */
+void
+InitClockAtBoot(void)
+{
+	if (creating_extension)
+	{
+		/* No catalog present yet */
+		return;
+	}
+
+	SpinLockAcquire(&logicalClockShmem->clockMutex);
+
+	/* Avoid repeated and parallel initialization */
+	if (logicalClockShmem->clockInitialized == INITIALIZED ||
+			logicalClockShmem->clockInitialized == IN_PROCESS)
+	{
+		SpinLockRelease(&logicalClockShmem->clockMutex);
+		return;
+	}
+
+	Assert(logicalClockShmem->clockInitialized == UN_INITIALIZED);
+
+	/*
+	 * Set the flag before executing a distributed query
+	 * (else it might trigger this routine recursively.)
+	 */
+	logicalClockShmem->clockInitialized = IN_PROCESS;
+
+	SpinLockRelease(&logicalClockShmem->clockMutex);
+
+	PG_TRY();
+	{
+		/* Start with the wall clock value */
+		uint64 epochValue = GetEpochTimeMs();
+		logicalClockShmem->clusterClockValue = START_NEW_CLOCK(epochValue);
+
+		/*
+		 * Version checking is a must for the below code as it accesses the
+		 * new catalog created in the new version.
+		 * Note: If we move this code to the top of the routine it will cause an
+		 * infinite loop with CitusHasBeenLoaded() calling this routine again.
+		 */
+		bool saveEnableVersionChecks = EnableVersionChecks;
+		EnableVersionChecks = true;
+
+		if (!CheckCitusVersion(NOTICE))
+		{
+			EnableVersionChecks = saveEnableVersionChecks;
+			return;
+		}
```

Comment:
if we return here we keep our `clockInitialized` status to `IN_PROGRESS` without actually progressing. This prevents any future progression I believe.

could anything go wrong here?

This mostly seems to apply to the backend that might CREATE the citus extension. It would be stuck with a non-initialized and never initializing counter.

Given this is in shared memory I wonder how it would ever work in testing and development without restarting the cluster. I might be missing something here.

actually looking at the implementation of `CheckCitusVersion` it seems to be returning true _always_ when the extension hasn't been created yet. I feel more and more inclined that this check might not be guarding against what you are wanting.

There are many areas of the C code where citus expects that the catalogs are up to date. Can you elaborate what this should be guarding against and why we are not using the general citus infra for it?

Reply:
>>if we return here we keep our clockInitialized status to IN_PROGRESS without actually progressing

Good catch, we should reset IN_PROGRESS. 

The ***only*** reason why we are doing this hack is, one of our internal upgrade test is explicitly setting this GUC to false and the test is failing.

Reply:
Can you link this upgrade tests?
It seems that it is a wrong use of the GUC in an upgrade test.

I think the only place where we change this GUC is in some consistency checks on our upgrade/downgrade scripts, because we _need_ to be able to query the database while the citus catalog and the binaries are on incompatible versions. This should be long after the clock has been initialized in that run.

Reply:
citus_tests/config.py:        self.new_settings = {"citus.enable_version_checks": "false"}

Discussion in our internal channel

![image](https://user-images.githubusercontent.com/44680808/192864090-c0658fd9-3667-4600-bd6c-9b493f496431.png)

Reply:
This routine became clunky for couple of reasons.

1) Upgrade test hack for version-checks
2) Routine is in itself re-entrant which can trigger recursion with CheckCitusVersion()

CitusHasBeenLoaded
  StartupCitusBackend
	StartupCitusBackend
		InitClockAtBoot

Reply:
OK, I've rearranged the code and removed the ugly version-check hack. It should be simpler now.